mesa/space.py

```diff
@@ -639,7 +639,7 @@ def iter_cell_list_contents(
             A iterator of the contents of the cells identified in cell_list
         """
         return itertools.chain.from_iterable(
-            self[x][y] for x, y in cell_list if not self.is_cell_empty((x, y))
+            self._grid[x][y] for x, y in cell_list if self._grid[x][y]
```

Comment:
Replacing `not self.is_cell_empty((x, y))` with this is nonlocal and problematic, because when one changes the implementation of `is_cell_empty`, this line needs to be changed as well, which is very easy to miss.

Reply:
this makes me think that we can instead of iterating as is we can apply a custom filter decided by the user, this could be a useful feature.

Reply:
For example:

from:

` return itertools.chain.from_iterable(self._grid[x][y] for x, y in cell_list if not self.is_cell_empty((x, y)))`

we can pass to


```
if exclude is None:
    return itertools.chain.from_iterable(self._grid[x][y] for x, y in cell_list)
else:
    return itertools.chain.from_iterable(self._grid[x][y] for x, y in cell_list if filter(exclude, cell_list))
```

Note that in the first case we have an iterator which returns non-empty content because it chains the content of the cells. What do you think?

Reply:
But the problem remains, in Grid we have an `iter_cell_list_contents` without reference to `is_cell_empty`, here we could do the same with my implementation, so what to do? I think that we can change the docstring to make it more clear that we are returning the agents in `cell_list` and this would be enough, something like `Returns an iterator of the agents in the
cells identified in cell_list.`

Reply:
What about now? (I changed also the annotation in  `MultiGrid.iter_cell_list_contents` which happened to be wrong)

Reply:
I think that you are right, so I made everything consistent.

Reply:
But right now I'm thinking that the strategy with the custom filter could be actually doable, because in any case the overhead created is just the same as using the default case right now and then filtering from the outside, so we would have the same speed in the custom case, but it would be faster in the default one, with also the added functionality

Reply:
@rht I think this one as it is now doesn't change the behaviour of the methods, improving them (2 in performance 1 in homogeneity with the other methods), and so could be merged as it is, the filter functionality (which I would like to hear your opinion about) can be done in another pr.